### PR TITLE
PR of just changes from desktop

### PR DESCRIPTION
**Pull request by twsouthwick**
This reverts commit a34c2022ff373bb4eb08211f45ae1471793cfd61.

---

This change is [Reviewable](https://reviewable.io/reviews/twsouthwick/corefx/1)

### CODE REVIEW

File: src/System.ComponentModel.TypeConverter/src/System/ComponentModel/AttributeCollection.cs [hunk 1]
```diff
@@ -396,28 +364,18 @@ public bool Matches(Attribute[] attributes)
         }
 
         /// <internalonly/>
-        int ICollection.Count 
+        bool ICollection.IsSynchronized
         {
```
Comment (Tanya-Solyanik) [inline src/System.ComponentModel.TypeConverter/src/System/ComponentModel/AttributeCollection.cs:368]: consider adding explicit "private" modifier for readability, here and everywhere else.

File: src/System.ComponentModel.TypeConverter/src/System/ComponentModel/EventHandlerList.cs [hunk 2]
```diff
@@ -27,40 +22,33 @@ public EventHandlerList()
...
         /// </devdoc>
-        public Delegate this[object key] {
+        public Delegate this[object key]
+        {
             [SuppressMessage("Microsoft.Security", "CA2122:DoNotIndirectlyExposeMethodsWithLinkDemands")]
-            get {
+            get
+            {
                 ListEntry e = null;
-                if (parent == null || parent.CanRaiseEventsInternal)
```
Comment (Tanya-Solyanik) [inline src/System.ComponentModel.TypeConverter/src/System/ComponentModel/EventHandlerList.cs:46]: Why are we not testing if parent can raise events any more?

File: src/System.ComponentModel.TypeConverter/src/System/ComponentModel/EventHandlerList.cs [hunk 3]
```diff
@@ -69,21 +57,25 @@ internal EventHandlerList(Component parent)
         ///    <para>[To be supplied.]</para>
```
Comment (Tanya-Solyanik) [inline src/System.ComponentModel.TypeConverter/src/System/ComponentModel/EventHandlerList.cs:57]: Consider: remove the useless comment

File: src/System.ComponentModel.TypeConverter/src/System/ComponentModel/GuidConverter.cs [hunk 4]
```diff
@@ -1,86 +1,87 @@
...
+        ///    <para>
+        ///        Gets a value indicating whether this converter can convert an object to
+        ///        the given destination type using the context.
+        ///    </para>
         /// </devdoc>
-        public override bool CanConvertTo(ITypeDescriptorContext context, Type destinationType) {
-            if (destinationType == typeof(InstanceDescriptor)) {
+        public override bool CanConvertTo(ITypeDescriptorContext context, Type destinationType)
+        {
+#if FEATURE_INSTANCEDESCRIPTOR
```
Comment (Tanya-Solyanik) [inline src/System.ComponentModel.TypeConverter/src/System/ComponentModel/GuidConverter.cs:38]: Why is this added as a conditional feature?

File: src/System.ComponentModel.TypeConverter/src/System/ComponentModel/MemberDescriptor.cs [hunk 5]
```diff
@@ -301,48 +321,60 @@ public abstract class MemberDescriptor {
...
+            if ((mdObj._category == null) != (_category == null) ||
+                (_category != null && !mdObj._category.Equals(_category)))
+            {
                 return false;
             }
 
-            if ((mdObj.description == null) != (description == null) ||
-                (description != null && !mdObj.category.Equals(description))) {
+            if ((mdObj._description == null) != (_description == null) ||
+                (_description != null && !mdObj._category.Equals(_description)))
```
Comment (Tanya-Solyanik) [inline src/System.ComponentModel.TypeConverter/src/System/ComponentModel/MemberDescriptor.cs:356]: category == description bug was fixed (qurked) in the desktop version, in 4.6.2. What branch did you use as a source? Or was this intentional?

File: src/System.ComponentModel.TypeConverter/src/System/ComponentModel/MemberDescriptor.cs [hunk 6]
```diff
@@ -358,48 +390,56 @@ public abstract class MemberDescriptor {
...
                         FillAttributes(list);
                     }
-                    catch (System.Threading.ThreadAbortException) {
-                        throw;
-                    }
-                    catch (Exception e) {
-                        Debug.Fail(name + ">>" + e.ToString()); 
+                    catch (Exception e)
+                    {
+                        Debug.Fail($"{_name}>>{e}");
```
Comment (Tanya-Solyanik) [inline src/System.ComponentModel.TypeConverter/src/System/ComponentModel/MemberDescriptor.cs:419]: Show off :)!

File: src/System.ComponentModel.TypeConverter/src/System/ComponentModel/MemberDescriptor.cs [hunk 7]
```diff
@@ -418,16 +459,27 @@ public abstract class MemberDescriptor {
...
+                // The original impementation requires the method https://msdn.microsoft.com/en-us/library/5fed8f59(v=vs.110).aspx which is not 
+                // available on .NET Core. The replacement will use the default BindingFlags, which may miss some methods that had been found
+                // on .NET Framework.
+#if FEATURE_GETMETHOD_WITHTYPES
+                result = componentClass.GetTypeInfo().GetMethod(name, BindingFlags.Instance | BindingFlags.Static | BindingFlags.Public | BindingFlags.NonPublic, null, args, null);
+#else
+                result = componentClass.GetTypeInfo().GetMethod(name, args);
+#endif
+            }
+            if (result != null && !result.ReturnType.GetTypeInfo().IsEquivalentTo(returnType))
```
Comment (Tanya-Solyanik) [inline src/System.ComponentModel.TypeConverter/src/System/ComponentModel/MemberDescriptor.cs:481]: Should not it be returnType.GetTypeInfo()?

File: src/System.ComponentModel.TypeConverter/src/System/ComponentModel/PropertyDescriptor.cs [hunk 8]
```diff
@@ -74,34 +65,39 @@ protected PropertyDescriptor(MemberDescriptor descr, Attribute[] attrs)
...
-                if (converter == null) {
+                if (_converter == null)
+                {
                     TypeConverterAttribute attr = (TypeConverterAttribute)attrs[typeof(TypeConverterAttribute)];
-                    if (attr.ConverterTypeName != null && attr.ConverterTypeName.Length > 0) {
+                    if (attr.ConverterTypeName != null && attr.ConverterTypeName.Length > 0)
+                    {
                         Type converterType = GetTypeFromName(attr.ConverterTypeName);
-                        if (converterType != null && typeof(TypeConverter).IsAssignableFrom(converterType)) 
+                        if (converterType != null && typeof(TypeConverter).GetTypeInfo().IsAssignableFrom(converterType))
```
Comment (Tanya-Solyanik) [inline src/System.ComponentModel.TypeConverter/src/System/ComponentModel/PropertyDescriptor.cs:89]: converterType.GetTypeInfo()?

File: src/System.ComponentModel.TypeConverter/src/System/ComponentModel/PropertyDescriptor.cs [hunk 9]
```diff
@@ -387,16 +419,17 @@ protected PropertyDescriptor(MemberDescriptor descr, Attribute[] attrs)
             // because the CLR may have cached an older version if the assembly's version number didn't change
             // See VSWhidbey 560732
             Type typeFromComponent = null;
-            if (ComponentType != null) {
+            if (ComponentType != null)
+            {
                 if ((typeFromGetType == null) ||
-                    (ComponentType.Assembly.FullName.Equals(typeFromGetType.Assembly.FullName))) {
-
+                    (ComponentType.GetTypeInfo().Assembly.FullName.Equals(typeFromGetType.GetTypeInfo().Assembly.FullName)))
```
Comment (Tanya-Solyanik) [inline src/System.ComponentModel.TypeConverter/src/System/ComponentModel/PropertyDescriptor.cs:425]: FxCOP would complain about the StringComparison argument.

File: src/System.ComponentModel.TypeConverter/src/System/ComponentModel/MemberDescriptor.cs (hunk 5 above)
Comment (chlowell) [inline src/System.ComponentModel.TypeConverter/src/System/ComponentModel/MemberDescriptor.cs:356]: @twsouthwick You can cherry-pick https://github.com/chlowell/corefx/commit/64b6b9fd6812dd20c98db371952e68caa49e8d4f to fix this

File: src/System.ComponentModel.TypeConverter/src/System/ComponentModel/AttributeCollection.cs (hunk 1 above)
Comment (twsouthwick) [inline src/System.ComponentModel.TypeConverter/src/System/ComponentModel/AttributeCollection.cs:368]: Explicit interface implementations don't allow visibility keywords

File: src/System.ComponentModel.TypeConverter/src/System/ComponentModel/MemberDescriptor.cs (hunk 5 above)
Comment (twsouthwick) [inline src/System.ComponentModel.TypeConverter/src/System/ComponentModel/MemberDescriptor.cs:356]: We used reference source as the source. We have the fix available, just haven't updated yet.

File: src/System.ComponentModel.TypeConverter/src/System/ComponentModel/MemberDescriptor.cs (hunk 7 above)
Comment (twsouthwick) [inline src/System.ComponentModel.TypeConverter/src/System/ComponentModel/MemberDescriptor.cs:481]: There's an overload that takes just a Type. We have a work item to go through and clean up reflection usage more, but don't want to make too many changes with an already too large PR :)

File: src/System.ComponentModel.TypeConverter/src/System/ComponentModel/MemberDescriptor.cs (hunk 5 above)
Comment (Tanya-Solyanik) [inline src/System.ComponentModel.TypeConverter/src/System/ComponentModel/MemberDescriptor.cs:356]: Consider diffing reference source against NetfxDev1 source, This is the only change that I know of in this namespace in 4.6.2, but maybe my records are incomplete...

File: src/System.ComponentModel.TypeConverter/src/System/ComponentModel/GuidConverter.cs (hunk 4 above)
Comment (twsouthwick) [inline src/System.ComponentModel.TypeConverter/src/System/ComponentModel/GuidConverter.cs:38]: There are a few features that we haven't enabled due to surface area not needing it. For now, we added conditional feature flags for them so the original source remains if we want to enable it later.

File: src/System.ComponentModel.TypeConverter/src/System/ComponentModel/EventHandlerList.cs (hunk 3 above)
Comment (twsouthwick) [inline src/System.ComponentModel.TypeConverter/src/System/ComponentModel/EventHandlerList.cs:57]: There's some work to include better documentation, so that will take care of it.

File: src/System.ComponentModel.TypeConverter/src/System/ComponentModel/EventHandlerList.cs (hunk 2 above)
Comment (chlowell) [inline src/System.ComponentModel.TypeConverter/src/System/ComponentModel/EventHandlerList.cs:46]: We removed `parent` entirely because it's a `Component` (unavailable in CoreFX).

File: src/System.ComponentModel.TypeConverter/src/System/ComponentModel/GuidConverter.cs (hunk 4 above)
Comment (chlowell) [inline src/System.ComponentModel.TypeConverter/src/System/ComponentModel/GuidConverter.cs:38]: `InstanceDescriptor` went behind a feature flag for now because typically it's used only like this, as a `CanConvertTo` special case.
